plugins/atomicfu/atomicfu-compiler/src/org.jetbrains.kotlinx.atomicfu.compiler/extensions/AtomicFUTransformer.kt

```diff
@@ -0,0 +1,379 @@
+    private val ATOMIC_ARRAY_TYPES = setOf("AtomicIntArray", "AtomicLongArray", "AtomicBooleanArray", "AtomicArray")
+    private val ATOMICFU_INLINE_FUNCTIONS = setOf("atomicfu_loop", "atomicfu_update", "atomicfu_getAndUpdate", "atomicfu_updateAndGet")
+
+    fun transform(irFile: IrFile) {
+        irFile.transform(AtomicExtensionTransformer(), null)
+        irFile.transformChildren(AtomicTransformer(), null)
+
+        irFile.patchDeclarationParents()
+    }
+
+    private inner class AtomicExtensionTransformer : IrElementTransformerVoid() {
+        override fun visitFile(declaration: IrFile): IrFile {
+            declaration.declarations.addAllTransformedAtomicExtensions()
+            return super.visitFile(declaration)
+        }
+
+        override fun visitClass(declaration: IrClass): IrStatement {
+            declaration.declarations.addAllTransformedAtomicExtensions()
+            return super.visitClass(declaration)
+        }
+
+        private fun MutableList<IrDeclaration>.addAllTransformedAtomicExtensions() {
+            val transformedDeclarations = mutableListOf<IrDeclaration>()
+            forEach { irDeclaration ->
+                irDeclaration.transformAtomicExtension()?.let { it -> transformedDeclarations.add(it) }
+            }
+            addAll(transformedDeclarations)
+        }
+
+        private fun IrDeclaration.transformAtomicExtension(): IrDeclaration? {
+            // Transform the signature of the inline Atomic* extension declaration:
+            // inline fun AtomicRef<T>.foo(arg) { ... } -> inline fun <T> foo(arg', atomicfu$getter: () -> T, atomicfu$setter: (T) -> Unit)
+            if (this is IrFunction && isAtomicExtension()) {
+                val newDeclaration = deepCopyWithSymbols(parent)
+                val valueParametersCount = valueParameters.size
+                val type = newDeclaration.extensionReceiverParameter!!.type.atomicToValueType()
+                val getterType = context.buildGetterType(type)
+                val setterType = context.buildSetterType(type)
+                newDeclaration.valueParameters = newDeclaration.valueParameters + listOf(
+                    buildValueParameter(newDeclaration, GETTER, valueParametersCount, getterType),
+                    buildValueParameter(newDeclaration, SETTER, valueParametersCount + 1, setterType)
+                )
+                newDeclaration.extensionReceiverParameter = null
+                return newDeclaration
+            }
+            return null
+        }
+    }
+
+    private inner class AtomicTransformer : IrElementTransformer<IrFunction?> {
+
+        override fun visitFunction(declaration: IrFunction, data: IrFunction?): IrStatement {
+            return super.visitFunction(declaration, declaration)
+        }
+
+        override fun visitCall(expression: IrCall, data: IrFunction?): IrElement {
+            expression.eraseAtomicFactory()?.let { return it.transform(this, data) }
```

Comment:
`expression.transformChildren(..)`

Reply:
Here I have to remove the atomic factory invocation at first and then transform its argument only. 
Consider, this example:
```
val a = atomic(b.value) // b is another atomic value
```
`eraseAtomicFactory` removes function call:
`expression.eraseAtomicFactory() == b.value` 
then we should transform this IrCall node `b.value` to just `b` and return as the result of transformation